DependencyInjection/SwiftmailerTransportFactory.php

```diff
@@ -0,0 +1,138 @@
+            $smtpAuthHandler->setAuthMode($options['auth_mode']);
+
+            $transport = new \Swift_Transport_EsmtpTransport(
+                new \Swift_Transport_StreamBuffer(new \Swift_StreamFilters_StringReplacementFilterFactory()),
+                array($smtpAuthHandler),
+                $eventDispatcher
+            );
+            $transport->setHost($options['host']);
+            $transport->setPort($options['port']);
+            $transport->setEncryption($options['encryption']);
+            $transport->setTimeout($options['timeout']);
+            $transport->setSourceIp($options['source_ip']);
+
+            (new SmtpTransportConfigurator(null, $requestContext))->configure($transport);
+        } elseif ('sendmail' === $transport) {
+            $transport = new \Swift_Transport_SendmailTransport(
+                new \Swift_Transport_StreamBuffer(new \Swift_StreamFilters_StringReplacementFilterFactory()),
+                $eventDispatcher
+            );
+
+            (new SmtpTransportConfigurator(null, $requestContext))->configure($transport);
+        } elseif ('mail' === $transport) {
+            $transport = new \Swift_Transport_MailTransport(
+                new \Swift_Transport_SimpleMailInvoker(),
+                $eventDispatcher
+            );
+        } elseif ('null' === $transport) {
+            $transport = new \Swift_Transport_NullTransport($eventDispatcher);
+        } else {
+            throw new \InvalidArgumentException(sprintf('Not a built-in Swiftmailer transport: %s.', $options['dsn']));
+        }
+
+        return $transport;
+    }
+
+    /**
+     * @param array $options
+     *
+     * @return string transport
+     */
+    public static function resolveOptions(array &$options)
+    {
+        if (null === $options['transport']) {
+            $transport = 'null';
+        } elseif ('gmail' === $options['transport']) {
+            $options['encryption'] = 'ssl';
+            $options['auth_mode'] = 'login';
+            $options['host'] = 'smtp.gmail.com';
+            $transport = 'smtp';
+        } else {
+            $transport = $options['transport'];
+        }
+
+        if (isset($options['dsn'])) {
+            $parts = parse_url($options['dsn']);
+            if (!empty($parts['scheme'])) {
+                $transport = $parts['scheme'];
+            }
+
+            if (!empty($parts['user'])) {
+                $options['username'] = $parts['user'];
+            }
+            if (!empty($parts['pass'])) {
+                $options['password'] = $parts['pass'];
+            }
+            if (!empty($parts['host'])) {
+                $options['host'] = $parts['host'];
+            }
+            if (!empty($parts['port'])) {
+                $options['port'] = $parts['port'];
+            }
+            if (!empty($parts['query'])) {
+                $query = [];
+                parse_str($parts['query'], $query);
+                if (!empty($query['encryption'])) {
```

Comment:
this function is missing null values for all options, in case the passed $options array does not contain all of them.
Then, this list of ifs could be replaced by a single foreach